File: backends/bot-backend/app/schemas/backtest.py
from pydantic import BaseModel, Field, validator
from typing import List, Optional, Dict, Any
from datetime import datetime

class BacktestCreate(BaseModel):
    """Initial request to start a backtest"""
    strategy_id: str = Field(..., description="Strategy identifier (e.g., 'MovingAverageCross')")
    name: str = Field(..., description="Human-readable name for this run")
    symbols: List[str] = Field(..., description="List of symbols to trade")
    interval: str = Field("1m", description="Candle timeframe")
    start_date: str = Field(..., description="Start date (ISO 8601)")
    end_date: str = Field(..., description="End date (ISO 8601)")
    initial_capital: float = Field(10000.0, gt=0)
    
    # Optional parameters
    strategy_params: Optional[Dict[str, Any]] = None
    risk_params: Optional[Dict[str, Any]] = None
    
    # Advanced settings (with defaults from config mostly, but overrideable)
    slippage_bps: Optional[float] = 10.0
    fee_bps: Optional[float] = 6.0
    market_type: str = "crypto"
    data_source: str = "binance"
    
    @validator("symbols")
    def validate_symbols(cls, v):
        if not v:
            raise ValueError("At least one symbol required")
        return [s.upper() for s in v]

class BacktestMetrics(BaseModel):
    """Performance metrics for a run"""
    total_trades: int = 0
    win_rate: float = 0.0
    net_pnl: float = 0.0
    gross_pnl: float = 0.0
    total_fees: float = 0.0
    max_drawdown: float = 0.0
    sharpe_ratio: Optional[float] = None
    return_pct: Optional[float] = None

class BacktestRun(BaseModel):
    """Backtest run details"""
    id: str
    user_id: str
    name: str
    strategy_id: str
    status: str
    created_at: str
    completed_at: Optional[str] = None
    
    # Config summary
    symbols: List[str]
    timeframe: str
    start_date: str
    end_date: str
    initial_capital: float
    
    # Result summary
    metrics: BacktestMetrics
    error_message: Optional[str] = None
    progress_pct: float = 0.0

class BacktestListResponse(BaseModel):
    """Paginated list of runs"""
    items: List[BacktestRun]
    total: int
    page: int
    size: int

class EquityPoint(BaseModel):
    """Single point in equity curve"""
    timestamp: str
    equity: float
    balance: float
    drawdown_pct: float
    unrealized_pnl: float

class EquityCurveResponse(BaseModel):
    """Full equity curve series"""
    run_id: str
    datapoints: List[EquityPoint]

class FillItem(BaseModel):
    """Single trade execution"""
    timestamp: str
    symbol: str
    side: str
    price: float
    quantity: float
    fee_usdt: float
    pnl: Optional[float] = None
    
class FillListResponse(BaseModel):
    """Paginated list of fills"""
    items: List[FillItem]
    total: int
    page: int
    size: int

File: frontends/user-frontend/src/api/backtest.ts
import { apiClient } from "./client";

export interface BacktestConfig {
    strategy_id: string;
    name: string;
    symbols: string[];
    interval: string;
    start_date: string;
    end_date: string;
    initial_capital: number;
    strategy_params?: Record<string, any>;
    risk_params?: Record<string, any>;
    slippage_bps?: number;
    fee_bps?: number;
    market_type?: string;
    data_source?: string;
}

export interface BacktestMetrics {
    total_trades: number;
    win_rate: number;
    net_pnl: number;
    gross_pnl: number;
    total_fees: number;
    max_drawdown: number;
    sharpe_ratio?: number;
    return_pct?: number;
}

export interface BacktestRun {
    id: string;
    user_id: string;
    name: string;
    strategy_id: string;
    status: "pending" | "processing" | "completed" | "failed" | "cancelled";
    created_at: string;
    completed_at?: string;
    symbols: string[];
    timeframe: string;
    start_date: string;
    end_date: string;
    initial_capital: number;
    metrics: BacktestMetrics;
    error_message?: string;
    progress_pct: number;
}

export interface BacktestListResponse {
    items: BacktestRun[];
    total: number;
    page: number;
    size: number;
}

export interface EquityPoint {
    timestamp: string;
    equity: number;
    balance: number;
    drawdown_pct: number;
    unrealized_pnl: number;
}

export interface EquityCurveResponse {
    run_id: string;
    datapoints: EquityPoint[];
}

export interface FillItem {
    timestamp: string;
    symbol: string;
    side: "BUY" | "SELL";
    price: number;
    quantity: number;
    fee_usdt: number;
    pnl?: number;
}

export interface FillListResponse {
    items: FillItem[];
    total: number;
    page: number;
    size: number;
}

export const BacktestAPI = {
    create: async (config: BacktestConfig): Promise<{ run_id: string; job_id: string }> => {
        const response = await apiClient.post("/api/v1/backtests/", config);
        return response.data;
    },

    list: async (params?: { status?: string; page?: number; size?: number }): Promise<BacktestListResponse> => {
        const response = await apiClient.get<BacktestListResponse>("/api/v1/backtests/", { params });
        return response.data;
    },

    get: async (runId: string): Promise<BacktestRun> => {
        const response = await apiClient.get<BacktestRun>(`/api/v1/backtests/${runId}`);
        return response.data;
    },

    getEquityCurve: async (runId: string): Promise<EquityCurveResponse> => {
        const response = await apiClient.get<EquityCurveResponse>(`/api/v1/backtests/${runId}/equity`);
        return response.data;
    },

    getFills: async (runId: string, params?: { page?: number; size?: number }): Promise<FillListResponse> => {
        const response = await apiClient.get<FillListResponse>(`/api/v1/backtests/${runId}/fills`, { params });
        return response.data;
    },

    cancel: async (runId: string): Promise<{ status: string; message: string }> => {
        const response = await apiClient.post(`/api/v1/backtests/${runId}/cancel`);
        return response.data;
    },

    getExportUrl: (runId: string, format: "csv" | "json" = "csv"): string => {
        const baseUrl = apiClient.defaults.baseURL || "";
        const token = localStorage.getItem("access_token");
        // Check if baseUrl ends with slash to avoid double slash, though usually fine
        const cleanBaseUrl = baseUrl.endsWith("/") ? baseUrl.slice(0, -1) : baseUrl;
        return `${cleanBaseUrl}/api/v1/backtests/${runId}/export?format=${format}&token=${token}`;
    }
};
